### DISCUSSION TITLE
SSO Login / Entra

### DISCUSSION THREAD

**Discussion by TroRon** [Stirling-Tools/Stirling-PDF#1952] [Q&A] [answered]
Have I missed something, is it not possible, or am I making a mistake somewhere?

Has anyone managed to restrict the use of Stirling PDF with Entra, or to authenticate the user via Entra?

If so, would anyone have a Docker Compose file as an example? 
The container always crashes completely for me.

I have managed to get the app to run, but after authentication I just end up back at the login screen. 

Currently the container doesn't even start anymore and I'm wondering why :(

Many thanks for a little help.

Ronny

**Comment by TroRon** [comment]
Does no one have any ideas on this?

**Comment by Frooodle** [comment]
Can you share your current settings?

**Reply by TroRon** [reply]
Sure.

Thanks for your reply and a time.

I have found many examples but want to hang the whole thing Entra, I'm sure I'm doing something wrong ...:
           SECURITY_OAUTH2_ENABLED: “true”
           SECURITY_OAUTH2_PROVIDER: M365
            SECURITY_OAUTH2_ISSUER: https://xxxx.....
            SECURITY_OAUTH2_CLIENTSECRET: xxxxxxx
            SECURITY_OAUTH2_SCOPES: openid,profile,email
            SECURITY_OAUTH2_USEASUSERNAME: email
            SECURITY_OAUTH2_AUTOCREATEUSER: “true”

**Comment by TroRon** [comment]
With this config, the container comes up:

services:
  stirling-pdf:
    container_name: test-pdf-tool
    image: frooodle/s-pdf:latest
    deploy:
      resources:
        limits:
          memory: 4G
    healthcheck:
      test:
        - CMD-SHELL
        - curl -f http://localhost:8080/api/v1/info/status | grep -q 'UP' &&
          curl -fL http://localhost:8080/ | grep -q 'Please sign in'
      interval: 5s
      timeout: 10s
      retries: 16
    ports:
      - 8075:8080
    volumes:
      - /docker/pdf_test/data:/usr/share/tessdata:rw
      - /docker/pdf_test/config:/configs:rw
      - /docker/pdf_test/logs:/logs:rw
    environment:
      DOCKER_ENABLE_SECURITY: "true"
      SECURITY_ENABLELOGIN: "true"
      SECURITY_OAUTH2_ENABLED: "true"
      SECURITY_OAUTH2_AUTOCREATEUSER: "true" # This is set to true to allow auto-creation of non-existing users in Stirling-PDF
      SECURITY_OAUTH2_ISSUER: https://accounts.google.com # Change with any other provider that supports OpenID Connect Discovery (/.well-known/openid-configuration) end-point
      SECURITY_OAUTH2_CLIENTID: <YOUR CLIENT ID>.apps.googleusercontent.com # Client ID from your provider
      SECURITY_OAUTH2_CLIENTSECRET: <YOUR CLIENT SECRET> # Client Secret from your provider
      SECURITY_OAUTH2_SCOPES: openid,profile,email # Expected OAuth2 Scope
      SECURITY_OAUTH2_USEASUSERNAME: email # Default is 'email'; custom fields can be used as the username
      SECURITY_OAUTH2_PROVIDER: AD Login (test Staff) # Set this to your OAuth provider's name, e.g., 'google' or 'keycloak'
      PUID: 1002
      PGID: 1002
      UMASK: "022"
      SYSTEM_DEFAULTLOCALE: de-DE
      UI_APPNAME: PDF Toolbox
      UI_HOMEDESCRIPTION: test - PDF Toolbox
      UI_APPNAMENAVBAR: test - PDF Toolbox
      SYSTEM_MAXFILESIZE: "100"
      METRICS_ENABLED: "true"
      SYSTEM_GOOGLEVISIBILITY: "true"
    restart: on-failure:5
networks: {}

However, as soon as I make adjustments for Azure, I get the following message:

The Issuer "https://sts.windows.net/xxxxxxxxxxxxxxxx/" provided in the configuration metadata did not match the requested issuer "https://login.microsoftonline.com/xxxxxxx/"

Does anyone here have any tips?

**Comment by TroRon** [comment]
Ok, I have now found out the Issuer URL at Azure...

Now the redirect URL is probably not yet correct ...

**Comment by TroRon** [comment, marked as answer]
Works now ;)

Preparations
************

Create an APP in Entra according to the known procedure.

Necessary information from the Entra app, after configuration:
**************************************************************

Issuer:
https://login.microsoftonline.com/TenantID/v2.0

Client ID
Client Secret

Reverse URL:
https://xxxxxxx.ch/login/oauth2/code/oidc



Working Config;
****************

services:
  stirling-pdf:
    container_name: Test-pdf-tool
    image: frooodle/s-pdf:latest
    deploy:
      resources:
        limits:
          memory: 4G
    healthcheck:
      test:
        - CMD-SHELL
        - curl -f http://localhost:8080/api/v1/info/status | grep -q 'UP' &&
          curl -fL http://localhost:8080/ | grep -q 'Please sign in'
      interval: 5s
      timeout: 10s
      retries: 16
    ports:
      - 8075:8080
    volumes:
      - /docker/pdf_Test/data:/usr/share/tessdata:rw
      - /docker/pdf_Test/config:/configs:rw
      - /docker/pdf_Test/logs:/logs:rw
    environment:
      DOCKER_ENABLE_SECURITY: "true"
      SECURITY_ENABLELOGIN: "true"
      SECURITY_OAUTH2_ENABLED: "true"
      SECURITY_OAUTH2_AUTOCREATEUSER: "true" # This is set to true to allow auto-creation of non-existing users in Stirling-PDF
      SECURITY_OAUTH2_ISSUER: Confidential # Change with any other provider that supports OpenID Connect Discovery (/.well-known/openid-configuration) end-point
      SECURITY_OAUTH2_CLIENTID: Confidential
      # Client ID from your provider
      SECURITY_OAUTH2_CLIENTSECRET: Confidential
      # Client Secret from your provider
      SECURITY_OAUTH2_SCOPES: openid,profile,email # Expected OAuth2 Scope
      SECURITY_OAUTH2_USEASUSERNAME: email # Default is 'email'; custom fields can be used as the username
      SECURITY_OAUTH2_PROVIDER: AD Login # Set this to your OAuth provider's name, e.g., 'google' or 'keycloak'
      PUID: 1002
      PGID: 1002
      UMASK: "022"
      SYSTEM_DEFAULTLOCALE: de-DE
      UI_APPNAME: PDF Toolbox
      UI_HOMEDESCRIPTION: Test - PDF Toolbox
      UI_APPNAMENAVBAR: Test - PDF Toolbox
      SYSTEM_MAXFILESIZE: "100"
      METRICS_ENABLED: "true"
      SYSTEM_GOOGLEVISIBILITY: "true"
    restart: on-failure:5
networks: {}

**Reply by Frooodle** [reply]
Any tips you can give to others with similar issues?

**Reply by hanwg** [reply]
I had issues configuring OIDC and was able to resolve it myself.
The value of SECURITY_OAUTH2_ISSUER should not end with "/.well-known/openid-configuration". This is because that suffix is automatically added.
Hope it's useful for others who might encounter the same.

**Reply by g33z** [reply]
Hey thanks for sharing all this - since I enabled OAuth2 today, found this through DuckDuckGo and found some (still) missing documentation, I thought I'd share the following:
Once you have enabled OAuth2, you might want to disable Form login, this can be done like so:

`SECURITY_LOGINMETHOD="oauth2"`

and can be found through the code in the discussion around here: https://github.com/Stirling-Tools/Stirling-PDF/pull/1471 but not in the docs, until today!
For I have created a pull request, adding this to places I would have been able to find it earlier, like the documentation! :stuck_out_tongue: 
https://github.com/Stirling-Tools/Stirling-Tools.github.io/pull/39

**Reply by Frooodle** [reply]
Thanks for updating doc, website should now contain latest changes 

Please update anything else you think might be missing or helpful!

**Reply by g33z** [reply]
> Please update anything else you think might be missing or helpful!

I'm on it! Thanks for the quick response!

**Comment by wmeneses** [comment]
Hi, thanks for this!!, I think my configuration is correct however I have this error and obviously docker does not start.
”
'org.springframework.security.config.annotation.web.configuration.WebSecurityConfiguration': Unsatisfied dependency expressed through field 'httpSecurity': Error creating bean with name 'org.springframework.security.config.annotation.web.configuration.HttpSecurityConfiguration': Unsatisfied dependency expressed through method 'setContentNegotiationStrategy' parameter 0: Error creating bean with name ”

**Reply by wmeneses** [reply]
My fault, I had left a space in a variable, it works fine!
